spec/spec.md

```diff
@@ -0,0 +1,2604 @@
+          out[0][0] <= in
+       else when and(eq(n, UInt(0)), eq(m, UInt(1))) :
+          out[0][1] <= in
+       else when and(eq(n, UInt(1)), eq(m, UInt(0))) :
+          out[1][0] <= in
+       else when and(eq(n, UInt(1)), eq(m, UInt(1))) :
+          out[1][1] <= in
+
+## Multiplexers
+
+A multiplexer outputs one of two input expressions depending on the
+value of an unsigned single bit selection signal.
+
+The following example connects to the `c` port the result of selecting
+between the `a` and `b` ports. The `a` port is selected when the `sel`
+signal is high, otherwise the `b` port is selected.
+
+    module MyModule :
+       input a: UInt
+       input b: UInt
+       input sel: UInt<1>
+       output c: UInt
+       c <= mux(sel, a, b)
+
+A multiplexer expression is legal only if the following holds.
+
+1.  The type of the selection signal is a single bit unsigned integer.
+
+2.  The types of the two input expressions are equivalent.
+
+3.  The types of the two input expressions are passive (see section
+    [4.4](#passive_types){reference-type="ref"
+    reference="passive_types"}).
+
+## Conditionally Valids {#conditionally_valids}
+
+A conditionally valid expression is expressed as an input expression
+guarded with an unsigned single bit valid signal. It outputs the input
+expression when the valid signal is high, otherwise the result is
+undefined.
+
+The following example connects the `a` port to the `c` port when the
+`valid` signal is high. Otherwise, the value of the `c` port is
+undefined.
+
+    module MyModule :
+       input a: UInt
+       input valid: UInt<1>
+       output c: UInt
+       c <= validif(valid, a)
+
+A conditionally valid expression is legal only if the following holds.
+
+1.  The type of the valid signal is a single bit unsigned integer.
+
+2.  The type of the input expression is passive (see section
+    [4.4](#passive_types){reference-type="ref"
+    reference="passive_types"}).
+
+Conditional statements can be equivalently expressed as multiplexers and
+conditionally valid expressions. See section
+[5.10](#conditionals) for
+details.
+
+## Primitive Operations
+
+All fundamental operations on ground types are expressed as a FIRRTL
+primitive operation. In general, each operation takes some number of
+argument expressions, along with some number of static integer literal
+parameters.
+
+The general form of a primitive operation is expressed as follows:
+
+    op(arg0, arg1, ..., argn, int0, int1, ..., intm)
+
+The following examples of primitive operations demonstrate adding two
+expressions, `a` and `b`, shifting expression `a` left by 3 bits,
+selecting the fourth bit through and including the seventh bit in the
+`a` expression, and interpreting the expression `x` as a Clock typed
+signal.
+
+    add(a, b)
+    shl(a, 3)
+    bits(a, 7, 4)
+    asClock(x)
+
+Section [7](#primitives)
+will describe the format and semantics of each primitive operation.
+
+# Primitive Operations {#primitives}
```

Comment:
all tables are broken.

Reply:
This is fixed in the latest version. E.g.:
<img width="923" alt="Screen Shot 2022-02-16 at 01 18 15" src="https://user-images.githubusercontent.com/1018530/154207578-8ec83c5d-e393-4e10-9735-a32974cbd9d0.png">

Reply:
Noting that some of the tables are overfull hboxes coming from tables, but this can be cleaned up later or by using larger margins for the spec.